### PR TITLE
Feature/22 provide mechanism distinguishing from multiple forms

### PR DESCRIPTION
**Pull request by pawelkoper**
---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java [hunk 1]
```diff
@@ -43,4 +43,10 @@ protected MultiMap fromJsonArray(final JsonArray array) {
 
         return map;
     }
+
+    protected String multiMapToString(MultiMap map) {
+        StringBuilder result = new StringBuilder();
+        map.names().forEach(header -> result.append(header).append("=").append(map.get(header)).append("\n"));
```
Comment (marcinczeczko) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java]: I think you can do it a little bita shorter :)
```java
return map.names().stream().reduce (new StringBuilder(), (x, y) -> x.append ("=").append(map.getHeader(y)).toString()
```

File: knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/service/ServiceEngine.java [hunk 2]
```diff
@@ -44,14 +47,20 @@ public ServiceEngine(Vertx vertx, TemplateEngineConfiguration serviceConfigurati
         this.configuration = serviceConfiguration;
     }
 
-    public Observable<Map<String, Object>> doServiceCall(ServiceEntry serviceEntry, MultiMap headers, HttpMethod httpMethod) {
+    public Observable<Map<String, Object>> doServiceCall(ServiceEntry serviceEntry, MultiMap headers, HttpMethod httpMethod, MultiMap formAttributes) {
```
Comment (marcinczeczko) [inline knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/service/ServiceEngine.java]: Well... I'd refactor this method a little bit, and use the approach from vert.x itself. THe observable created here suppose to move to kind of helper class, let's call it KnotxRxHelper in which we can but a metod for doing request that returns observable, in the form as below (it's just a draft - please double check what I have written )
```
    private Observable<HttpClientResponse> request(HttpClient client, HttpMethod method, int port, String domain, String uri, Action1<HttpClientRequest> requestBuilder) {
        return Observable.create(subscriber -> {
            HttpClientRequest req = client.request(method, port, domain, uri);
            requestBuilder.call(req);
            Observable<HttpClientResponse> resp = req.toObservable();
            resp.subscribe(subscriber);
            req.end();
        });
    }
```
Last parameter is the function call so we can use it to actually compose the request appropriately. E.g. we can invoke our helper:
```
        Observable<HttpClientResponse> serviceResponse = KnotxRxHelper.request(vertx.createHttpClient(), httpMethod, serviceEntry.getPort(), serviceEntry.getDomain(), serviceEntry.getServiceUri(),
                req -> buildRequestBody(req, headers, formAttributes));
```
While buildRequestBody() method is:
```
    private void buildRequestBody(HttpClientRequest request, MultiMap headers, MultiMap formAttributes) {
        request.headers().addAll(headers);
        if (!formAttributes.isEmpty()) {
            Buffer buffer = createFormPostBody(formAttributes);
            request.headers().set("content-length", String.valueOf(buffer.length()));
            request.headers().set("content-type", "application/x-www-form-urlencoded");
            request.write(buffer);
        }
```

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java (hunk 1 above)
Comment (pawelkoper) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java]: I've decided to use Guava's joiner instead

File: knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/service/ServiceEngine.java (hunk 2 above)
Comment (pawelkoper) [inline knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/service/ServiceEngine.java]: Thanks. I've refactor this code based on your comment

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java [hunk 3]
```diff
@@ -43,4 +47,8 @@ protected MultiMap fromJsonArray(final JsonArray array) {
 
         return map;
     }
+
+    protected String multiMapToString(MultiMap map) {
+        return Joiner.on("\n").withKeyValueSeparator("=").join((Iterable<Map.Entry<String, String>>) map.getDelegate());
```
Comment (marcinczeczko) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java]: looks like there is missing depednency to guava now as travis is failing on this while compiling

Comment (pawelkoper) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java]: There is  guava dependency in knotx-core/knotx-api/pom.xml. I will investigate why travis build is falling

Comment (pawelkoper) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/JsonObjectRequest.java]: it works now

**Comment by marcinczeczko** [comment]
PR approved. However is there any documentation update needs to be done for the changes introduced here ?

**Comment by marcinczeczko** [comment]
Approved. Please merge

**Pull request merged**